Translation: "Error Boundaries" (#108)
reactjs/de.reactjs.org: content/docs/error-boundaries.md

---
id: error-boundaries
title: Fehlergrenzen
permalink: docs/error-boundaries.html
---

> In der Vergangenheit, führten JavaScript-Fehler in Komponenten zu einem fehlerhaften Zustand innerhalb von React, welcher ein [Auftreten](https://github.com/facebook/react/issues/4026) [kryptischer](https://github.com/facebook/react/issues/6895) [Fehlermeldungen](https://github.com/facebook/react/issues/8579) in den folgenden Render-Vorgängen verursachte. Diese Fehler wurden immer von einem zuvor aufgetretenen Fehler in der Applikation verursacht, jedoch war seitens React keine Möglichkeit bereitgestellt, um auf diese Fehler innerhalb der Komponenten zu reagieren und sich von diesen zu erholen.

Author: Should I rather use "Anwendung" instead of "Applikation"? (Consider this for all occurrences of "Applikation")

Reviewer: Use "Anwendung". I think that's a better translation than "Applikation".

> In der Vergangenheit, führten JavaScript-Fehler in Komponenten zu einem fehlerhaften Zustand innerhalb von React, welcher ein [Auftreten](https://github.com/facebook/react/issues/4026) [kryptischer](https://github.com/facebook/react/issues/6895) [Fehlermeldungen](https://github.com/facebook/react/issues/8579) in den folgenden Render-Vorgängen verursachte. Diese Fehler wurden immer von einem zuvor aufgetretenen Fehler in der Applikation verursacht, jedoch war seitens React keine Möglichkeit bereitgestellt, um auf diese Fehler innerhalb der Komponenten zu reagieren und sich von diesen zu erholen.

Reviewer 2: "um mit Fehlern in Komponenten elegant umzugehen"

## Einführung zu Fehlergrenzen {#introducing-error-boundaries}

Ein JavaScript-Fehler in einem Teil der UI sollte nicht die ganze Applikation zerstören. Um eine Lösung für dieses Problem bereitzustellen, wurde mit React 16 das Konzept einer "Fehlergrenze" vorgestellt.

> Fehlergrenzen sind React-Komponenten die **JavaScript-Fehler im ihrem ganzen Kind-Komponenten Baum abfangen, diese loggen und eine Fallback-UI**, anstatt eines zerstörten Komponenten-Baums anzeigen. Fehlergrenzen fangen Fehler während des Renderings, in Lifecycle-Methoden und in Konstruktoren des ganzen Komponenten-Baums darunter.

Reviewer 2: in ihrem ganzen

## Live Demo {#live-demo}

Schau dir [folgendes Beispiel für die Deklaration und Nutzung einer Fehlergrenze](https://codepen.io/gaearon/pen/wqvxGa?editors=0010) mit [React 16](/blog/2017/09/26/react-v16.0.html) an.

## Wo platziert man Fehlergrenzen {#where-to-place-error-boundaries}

Über die Granularität der Fehlergrenzen kannst du frei entscheiden. Du kannst die top-level Route-Komponenten umschließen, um dem User eine "Etwas ist schiefgelaufen" Meldung anzuzeigen, diese Variante wird oft von serverseitigen Frameworks zur Fehlerbehandlung eingesetzt. Du kannst aber auch individuelle Widgets mit einer Fehlergrenze umschließen, um zu verhindern, dass diese den Rest der Anwendung zum Absturz bringen.

## Neues Verhalten für nicht abgefangene Fehlermeldungen {#new-behavior-for-uncaught-errors}

Diese Änderung hat eine wesentliche Auswirkung. **Mit React 16 führen Fehler, die nicht von einer Fehlergrenze abgefangen wurden, zum Unmounten des gesamten React Komponenten-Baums.**

> Wir haben über diese Entscheidung diskutiert, unserer Erfahrung nach ist es jedoch schlimmer eine fehlerhafte UI anzuzeigen, anstatt diese komplett zu entfernen. Zum Beispiel in einem Produkt wie ein Messenger, wäre es möglich, dass durch das Anzeigen einer fehlerhaften UI, die Nachricht an eine falsche Person verschickt werden könnte. Ähnlich schlimm wäre es in einer Bezahlanwendung, einen falschen Betrag, statt einfach nichts zu rendern.

Reviewer: ```suggestion
Wir haben über diese Entscheidung diskutiert, unserer Erfahrung nach ist es jedoch schlimmer eine fehlerhafte UI anzuzeigen, anstatt diese komplett zu entfernen. Zum Beispiel in einem Produkt wie einem Messenger, wäre es möglich, dass durch das Anzeigen einer fehlerhaften UI, die Nachricht an eine falsche Person verschickt werden könnte. Ebenso ist es in einer Zahlunganwendung schlimmer einen falschen Betrag anzuzeigen, statt einfach nichts zu rendern.
```

}
```

React-Komponenten sind deklarativ und spezifizieren *was* gerendert werden soll:

```js
<Button />
```

Fehlergrenzen bewahren die deklarative Natur von React und verhalten sich so wie du es erwarten würdest. Zum Beispiel, sogar wenn ein Fehler in einer `componentDidUpdate` Methode auftritt, verursacht von einem `setState` irgendwo tief im Komponenten-Baum, wird dieser Fehler trotzdem zur nächstgelegenen Fehlergrenze weitergeleitet.

## Wie sieht es mit Event-Handler aus? {#how-about-event-handlers}

Fehlergrenzen fangen **keine** Fehler innerhalb der Event-Handler ab.

> React braucht keine Fehlergrenzen um sich von Fehlern innerhalb von Event-Handler zu erholen. Im Gegensatz zu Render- und Lifecycle-Methoden finden Event-Handler nicht während des Renderings statt. Das heißt, im Falle eines Fehlers weiß React noch immer, was angezeigt werden soll.

Reviewer: ```suggestion
React braucht keine Fehlergrenzen um sich von Fehlern innerhalb von Event-Handlern zu erholen. Im Gegensatz zu Render- und Lifecycle-Methoden finden Event-Handler nicht während des Renderings statt. Das heißt, im Falle eines Fehlers weiß React noch immer, was angezeigt werden soll.
```

> Wenn du einen Fehler innerhalb von Event-Handler abfangen möchtest, nutze das normale `try` / `catch` JavaScript-Statement:

Reviewer: ```suggestion
Wenn du einen Fehler innerhalb von Event-Handlern abfangen möchtest, nutze das normale `try` / `catch` JavaScript-Statement:
```
